File: 프로그래머스/알고리즘 고득점 Kit/깊이,너비 우선 탐색(DFS,BFS)/Level 3/단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <map>
#include <queue>
using namespace std;

bool possible(string s1, string s2){
    int cnt = 0;
    for(int i=0; i<s1.length(); i++)
        if(s1[i] != s2[i]) cnt++;
    
    return cnt==1 ? true : false;
}

int solution(string begin, string target, vector<string> words) {
    map<string, int> visit;    // map[str]: str 되기 위해 몇 단계 거쳤는지
    for(string word : words) visit[word] = -1;
    visit[begin] = 0;
    
    queue<string> q;
    q.push(begin);
    
    while(q.size()){
        string tmp = q.front();
        q.pop();
        
        for(string word : words){
            if(visit[word]==-1 && possible(tmp, word)){
                visit[word] = visit[tmp]+1;
                q.push(word);
            }
        }
    }
    
    return visit[target]!=-1 ? visit[target] : 0;
}

File: 프로그래머스/알고리즘 고득점 Kit/깊이,너비 우선 탐색(DFS,BFS)/Level 3/단어 변환/단어 변환.py
def possible(s1, s2):
    cnt = 0
    for i in range(len(s1)):
        if s1[i] != s2[i]:
            cnt += 1

    if cnt == 1: return True
    else: return False

def solution(begin, target, words):
    # words 안에 target이 없는 경우
    if target not in words: return 0

    visit = {}
    for word in words:
        visit[word] = -1
    visit[begin] = 0

    q = [begin]

    while len(q) > 0:
        tmp = q[0]
        q.pop(0)

        for word in words:
            if visit[word]==-1 and possible(tmp, word):
                visit[word] = visit[tmp]+1
                q.append(word)

    if visit[target] != -1: return visit[target]
    else: return 0
